File: 백준/Silver/1629. 곱셈/곱셈.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

function solution(input) {
    let [A, B, C] = input[0].split(' ').map(BigInt);

    function powerMod(A, B, C) {
        if (B === BigInt(0)) return BigInt(1);
        if (B === BigInt(1)) return A % C;

        let result = powerMod(A, B / BigInt(2), C);
        result = (result * result) % C;

        if (B % BigInt(2) === BigInt(1)) return (result * A) % C;

        return result;
    }

    return powerMod(A, B, C).toString();
}

console.log(solution(input));

File: 백준/Silver/1629. 곱셈/곱셈.py
a, b, c = map(int, input().split())


def func1(a, b, m):
    # Base condition
    # a 가 m 보다 클 수 있으므로 a % m
    if (b == 1):
        return a % m

    val = func1(a, b//2, m)
    val = (val * val) % m

    # 짝수
    if (b % 2 == 0):
        return val

    # 홀수
    return val * a % m


print(func1(a, b, c))
